### PR TITLE
feat(core): add a feature to pick date range in month year format

### LINKED ISSUE #12319: pick Year/Month range in range date picker
#### Is this a bug, enhancement, or feature request?
enhancement

#### Describe your proposal.
examples [here](https://ui5.sap.com/1.108.34/#/entity/sap.m.DateRangeSelection/sample/sap.m.sample.DateRangeSelection)
only for range date picker

### PR DESCRIPTION
**Pull request by shawchandankumar**
Add a feature to pick date range in a month year format in date-picker component.

Closes  #12319

### CODE REVIEW

File: libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts [hunk 1]
```diff
@@ -213,16 +262,183 @@ export class CalendarMonthViewComponent<D> implements OnInit, OnChanges, Focusab
...
+        const cellToFocus = new DefaultCalendarActiveCellStrategy().getActiveCell(this._getMonthList());
+        if (!cellToFocus?.id) {
+            return;
+        }
+        this._focusElementBySelector(`#${cellToFocus.id}`);
+    }
+
+    /**
+     * @hidden
+     * Handles hover effect for single-range selection mode.
```
Comment (mikerodonnell89) [inline libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts:295]: What is "single-range"? I think this should just be "range"?

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
Good stuff, I left one comment above, also:

- I think to satisfy the requirements of the original issue, we also need the year-only selection too
- Range hover should work with keyboard as well

File: libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts [hunk 2]
```diff
@@ -54,15 +56,30 @@ export class CalendarMonthViewComponent<D> implements OnInit, OnChanges, Focusab
...
+    @Input()
+    isDateRangeMonthYearFormat: boolean;
+
     /** An event fired when a new month is selected */
     @Output()
     readonly monthClicked: EventEmitter<number> = new EventEmitter<number>();
 
+    /** Event thrown every time selected first or last date in range mode is changed */
+    @Output()
+    readonly selectedRangeDateChange: EventEmitter<DateRange<D>> = new EventEmitter<DateRange<D>>();
```
Comment (dpavlenishvili) [inline libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts:69]: why you need selectedRangeDateChange for month selection?

File: libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts [hunk 3]
```diff
@@ -54,15 +56,30 @@ export class CalendarMonthViewComponent<D> implements OnInit, OnChanges, Focusab
...
+
+    /**
+     * Whether user wants to mark day cells on hover.
```
Comment (dpavlenishvili) [inline libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts:72]: according this feature, user want to select mont or year not a day.

File: libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts [hunk 4]
```diff
@@ -82,6 +99,11 @@ export class CalendarMonthViewComponent<D> implements OnInit, OnChanges, Focusab
     /** @hidden */
     private _initiated = false;
 
+    /** @hidden */
+    private _selectedRangeDate: DateRange<D>;
```
Comment (dpavlenishvili) [inline libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts:103]: same thing, why selectedRangeDate

**Review by dpavlenishvili — CHANGES_REQUESTED** [review]
i left one general and other comments, check it and provide answer or make a code update

File: libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts (hunk 3 above)
Comment (shawchandankumar) [reply libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts:72]: @dpavlenishvili to notify the calendar component for a month change or newly month selected

File: libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts (hunk 4 above)
Comment (shawchandankumar) [reply libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts:103]: @dpavlenishvili To store the selectedDate locally in the component and to apply hover styles on the range

File: libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts (hunk 2 above)
Comment (shawchandankumar) [reply libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts:69]: @dpavlenishvili there is no specific way to store only the month/year so doing this

File: libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts (hunk 1 above)
Comment (shawchandankumar) [reply libs/core/calendar/calendar-views/calendar-month-view/calendar-month-view.component.ts]: @mikerodonnell89 corrected

**Comment by dpavlenishvili** [comment]
Range hover is not working well, if you select any year or month and then use left and right arrows, it starts hovering from different index. to reproduce try to select any new month or year, don't move mouse and use arrow keys.
image

**Comment by shawchandankumar** [comment]
> Range hover is not working well, if you select any year or month and then use left and right arrows, it starts hovering from different index. to reproduce try to select any new month or year, don't move mouse and use arrow keys. image

@dpavlenishvili It's only happening for the year selection not for the month selection.
Fixed the issue for the year range hover

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
There's also this problem with the hovering where the size of the cells change if they have the hover effect, and this causes the entire popover to change size. 


https://github.com/user-attachments/assets/eb1ae231-40b6-47c4-a877-991825a49b00

Should adopt how core calendar styles these hovered cells, see here: https://fundamental-ngx.netlify.app/#/core/calendar#range-hover

**Review by dpavlenishvili — APPROVED** [review]

**Comment by dpavlenishvili** [comment]
there is failing one unit test, @shawchandankumar please check it and fix it.

**Review by dpavlenishvili — CHANGES_REQUESTED** [review]
failing test

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
I think this still needs quite a bit of styles work:


https://github.com/user-attachments/assets/2fdc0fed-99fb-45e6-af23-003f2ad44c32

In the above video:
* weird focus on "September" when navigating with arrow keys
* Item sizes still grow vertically when they have the range style
* Double focus on "September" when navigating with mouse

**Comment by shawchandankumar** [comment]
> I think this still needs quite a bit of styles work:
> 
>  Screen.Recording.2024-09-18.at.11.59.57.AM.mov 
> In the above video:
> 
> * weird focus on "September" when navigating with arrow keys
> * Item sizes still grow vertically when they have the range style
> * Double focus on "September" when navigating with mouse

@mikerodonnell89 I think this styling part has to be corrected from the fundamental-styles side

**Pull request merged**